### PR TITLE
Implement permission levels for campaigns

### PR DESCRIPTION
**Pull request by mrnugget**
This fixes https://github.com/sourcegraph/sourcegraph/issues/10808 by implementing permission levels for campaigns. As described in that ticket there are two permission levels:

- `campaign-admin-level`: can update/close/delete/sync/... a campaign
- `non-admin-level`: can read the campaign

Right now, the a user has `campaign-admin-level` permissions if:
- the user is the author of the campaign
- OR is a site-admin.

What this PR does is to check for `campaign-admin-level` permissions in these mutations:

- `publishCampaign`
- `publishChangeset`
- `retryCampaign`
- `updateCampaign`
- `addChangesetsToCampaign`
- `syncChangeset`
- `closeCampaign`
- `deleteCampaign`

If the current user doesn't have `campaign-admin-level` permissions, an error is returned.

On the read-path it hides the errors in `campaign.status.errors` for users with non-admin-level permissions and makes sure that `campaign.viewerCanAdminister` returns the correct value.

### IMPORTANT: can this PR be merged without repository permissions?

Yes. Right now all top-level read-paths (`CampaignByID`, `Campaigns`, `ChangesetByID`, ...) and the mutations that allow creating new entities (`CreateCampaign`, `CreatePatchSet`) are still protected by our "campaigns are only available to site-admins"-check.

That will be removed in #10713 which we aim to implement after adding repository permissions to campaigns.

That means **it's safe to merge this PR**, even though we don't have repository permissions in place yet, **because only site-admins can create campaigns and thus become campaign owners**.

### Implementation

In order to be able to test the checks consistently I moved all of them into `campaigns.Service` along with business logic that was previously only defined on the resolver level. That code was previously completely untested and its functionality is still untested (since I think it's out of the scope of this PR), but at least we now check for the authorization level and execute the code in the tests.

### CODE REVIEW

**Comment by mrnugget** [comment]
Sorry, reviewers, had to force push because I needed to rebase onto master.

File: enterprise/internal/campaigns/resolvers/resolver.go [hunk 1]
```diff
@@ -391,14 +341,11 @@ func (r *Resolver) RetryCampaign(ctx context.Context, args *graphqlbackend.Retry
...
 
-	campaign, err := r.store.GetCampaign(ctx, ee.GetCampaignOpts{ID: campaignID})
-	if err != nil {
-		return nil, errors.Wrap(err, "getting campaign")
-	}
-
-	err = r.store.ResetFailedChangesetJobs(ctx, campaign.ID)
+	svc := ee.NewService(r.store, r.httpFactory)
+	// 🚨 SECURITY: RetryPublishCampaign checks whether current user is authorized.
+	campaign, err := svc.RetryPublishCampaign(ctx, campaignID)
```
Comment (eseliger) [inline enterprise/internal/campaigns/resolvers/resolver.go:346]: It's a bit hard to grasp for resolvers that lack authorization here, but I don't know any better way either

Comment (mrnugget) [reply enterprise/internal/campaigns/resolvers/resolver.go:346]: That's why I added the comments, to make that explicit.

In other parts of the codebase we have these permission checks strewn about in the resolvers. I've debated with myself (and @ryanslade) as to where to put these checks and Ryan ultimately convinced me that the right place is in the service, where we have all of our business logic.

File: enterprise/internal/campaigns/service.go [hunk 2]
```diff
@@ -392,6 +412,130 @@ func (s *Service) CloseOpenChangesets(ctx context.Context, cs []*campaigns.Chang
...
+	if _, err := s.store.GetChangeset(ctx, GetChangesetOpts{ID: id}); err != nil {
+		return err
+	}
+
+	campaigns, _, err := s.store.ListCampaigns(ctx, ListCampaignsOpts{ChangesetID: id})
+	if err != nil {
+		return err
+	}
+
+	// Check whether the user has admin rights for one of the campaigns.
```
Comment (ryanslade) [inline enterprise/internal/campaigns/service.go:485]: nit:

```suggestion
	// Check whether the user has admin rights for any of the campaigns.
```

File: enterprise/internal/campaigns/service.go [hunk 3]
```diff
@@ -392,6 +412,130 @@ func (s *Service) CloseOpenChangesets(ctx context.Context, cs []*campaigns.Chang
...
+	}
+
+	if err := repoupdater.DefaultClient.EnqueueChangesetSync(ctx, []int64{id}); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// RetryPublishCampaign resets the failed (!) ChangesetJobs for the given
```
Comment (ryanslade) [inline enterprise/internal/campaigns/service.go:512]: Why the `(!)`?

File: enterprise/internal/campaigns/service.go [hunk 4]
```diff
@@ -392,6 +412,130 @@ func (s *Service) CloseOpenChangesets(ctx context.Context, cs []*campaigns.Chang
...
+// campaign, which causes them to be re-run in the background.
+func (s *Service) RetryPublishCampaign(ctx context.Context, id int64) (campaign *campaigns.Campaign, err error) {
+	traceTitle := fmt.Sprintf("campaign: %d", id)
```
Comment (ryanslade) [inline enterprise/internal/campaigns/service.go:515]: Why do we have tracing here and not in the other methods?

File: enterprise/internal/campaigns/service.go (hunk 3 above)
Comment (mrnugget) [reply enterprise/internal/campaigns/service.go:512]: Just to make clear that it only resets the failed ChangesetJobs, as opposed to succeeded or something.

File: enterprise/internal/campaigns/service.go (hunk 4 above)
Comment (mrnugget) [reply enterprise/internal/campaigns/service.go:515]: Woops! Forgot the others. Will add tracing there too.

**Pull request merged**